pyqtorch/noise.py

```diff
@@ -366,4 +387,55 @@ def __init__(
             [[0, 0], [sqrt(rate), 0]], dtype=DEFAULT_MATRIX_DTYPE
         )
         kraus_generalized_amplitude_damping: list[Tensor] = [K0, K1, K2, K3]
-        super().__init__(kraus_generalized_amplitude_damping, target, probability)
+        super().__init__(kraus_generalized_amplitude_damping, target, error_probability)
+
+
+class NoiseProtocol:
+    BITFLIP = "BitFlip"
+    PHASEFLIP = "PhaseFlip"
+    DEPOLARIZING = "Depolarizing"
+    PAULI_CHANNEL = "PauliChannel"
+    AMPLITUDE_DAMPING = "AmplitudeDamping"
+    PHASE_DAMPING = "PhaseDamping"
+    GENERALIZED_AMPLITUDE_DAMPING = "GeneralizedAmplitudeDamping"
+
+    def __init__(self, protocol: str, options: dict = dict()) -> None:
```

Comment:
Not sure I understand what you mean. I think this argument can only be a string?

Reply:
Yeah sorry for being unclear. Maybe something like this:

```python
class NoiseProtocol:
    BITFLIP = "BitFlip"
    PHASEFLIP = "PhaseFlip"
    DEPOLARIZING = "Depolarizing"
    PAULI_CHANNEL = "PauliChannel"
    AMPLITUDE_DAMPING = "AmplitudeDamping"
    PHASE_DAMPING = "PhaseDamping"
    GENERALIZED_AMPLITUDE_DAMPING = "GeneralizedAmplitudeDamping"

    def __init__(self, protocols: str | list, options: dict = dict()) -> None:
         self.protocols: list = protocols if list else [protocols]
        self.options: dict = options
```

Reply:
Essentially capture the fact that you could have a sequence of noisy protocols in `NoiseProtocol` so that you only have to annotate using `noise: NoiseProtocol | None` in other classes.

Reply:
Ah but I don't think this is the current logic. From what I see in the `QuantumOperation.forward`, the noise is either a single `NoiseProtocol` or a `dict` with various `NoiseProtocol`. But for each `NoiseProtocol`, the `__init__` only takes a string, I don't see any logic to handle a list of strings.

I would revisit this logic in a future MR.

Reply:
Yeah the idea is to enable applying a sequence of these protocols. OK to leave it for now but shouldn't we remove the dicts of `NoiseProtocol` then ? Because my understanding was that it was implemented for this case.

Reply:
Yes ok I see your point. Definitely it would be cleaner I think. I've opened an issue here: https://github.com/pasqal-io/pyqtorch/issues/252